### PR TITLE
accept-headers: Support preference ordering

### PR DESCRIPTION
**Pull request by endocrimes**
Fixes content negotiation for Content Encodings - Namely supports parsing preferences `;q=FLOAT`, and resolving across multiple headers. Inspired by #55.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -355,72 +465,65 @@ mod tests {
     use super::*;
     use http::header::ACCEPT_ENCODING;
     use scopeguard::defer;
-    use std::{fs, path::Path, sync::Mutex};
-
-    static TEST_MUTEX: Mutex<()> = Mutex::new(());
```
Comment (endocrimes) [inline src/lib.rs:360]: I can re-add the test mutex if desired - but it made test failures much harder to understand by polluting output with `Poison {}` errors

Comment (itowlson) [reply src/lib.rs:360]: cc @dicej who added `TEST_MUTEX` as part of #43 and will have a better perspective than me on the impact of removing it

File: src/lib.rs [hunk 2]
```diff
@@ -49,47 +53,151 @@ const FALLBACK_FAVICON_ICO: &[u8] = include_bytes!("../spin-favicon.ico");
...
+#[derive(PartialEq)]
+struct ContentEncoding {
+    // We limit expressed encodings to ones that we support
+    encoding: SupportedEncoding,
+    weight: Option<f32>,
+}
+
+impl fmt::Display for ContentEncoding {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self.encoding)?;
```
Comment (itowlson) [inline src/lib.rs:65]: I'm always a bit wary of using a debug format in a Display implementation.  How would you feel about implementing `Display` on `SupportedEncoding`?

Looking at the Debug derivation on SupportedEncoding, this looks the Display implementationm will be different from the consts we use for parsing (e.g. `Brotli` rather than `br`) but 1. I may be misunderstanding and 2. we may not care - just flagging it just in case.

File: src/lib.rs [hunk 3]
```diff
@@ -49,47 +53,151 @@ const FALLBACK_FAVICON_ICO: &[u8] = include_bytes!("../spin-favicon.ico");
...
+                // Otherwise prefer the more specific option
+                (SupportedEncoding::None, _) => Some(Ordering::Less),
+                (_, SupportedEncoding::None) => Some(Ordering::Greater),
+                // Everything else is roughly equal
+                (_, _) => Some(Ordering::Equal),
+            },
+            v => v,
+        }
+    }
+}
```
Comment (itowlson) [inline src/lib.rs:105]: The amount of encoding-specific stuff makes me think it's time to move it to an `encoding` module and keep it out of the main flow.  But appreciate that having it here makes it easier to see what's changed!

File: src/lib.rs [hunk 4]
```diff
@@ -49,47 +53,151 @@ const FALLBACK_FAVICON_ICO: &[u8] = include_bytes!("../spin-favicon.ico");
...
-                                    (v.trim().to_lowercase() == *name).then_some(*encoding)
-                                })
-                            })
+            .filter(|(k, _)| HeaderName::from_bytes(k.as_bytes()).ok() == Some(ACCEPT_ENCODING))
+            .flat_map(|(_, v)| {
+                str::from_utf8(v).ok().map(|v| {
+                    v.split(',')
+                        .map(|v| ContentEncoding::from_str(v).ok())
+                        .filter(|v| match v {
+                            Some(y) => match y.encoding {
```
Comment (itowlson) [inline src/lib.rs:171]: Might be able to replace this with `!matches!(y, SupportedEncoding::None)` if you want (I am not sure if it's clearer though!)

Or even replace the `match v` on the previous line with `!matches(v, None | Some(SupportedEncoding::None))`

**Review by itowlson — APPROVED** [review]
I had a couple of minor suggestions but nothing blocking.  LGTM though it would be great if you could hold off until Joel can comment on the mutex.  Thanks!

File: src/lib.rs (hunk 1 above)
Comment (dicej) [reply src/lib.rs:360]: The reason I added the mutex is that all of the tests either read or write (or both) the same environment variables, and since `cargo test` defaults to concurrent test execution, the tests will spuriously interfere with each other and fail non-deterministically unless e.g. `cargo test -- --test-threads=1` is specified.  It took me a while to figure that out, and I didn't want anyone else to have to do it also.

In short: we should definitely either:
- keep the mutex
- come up with a different way to serialize the tests by default
- figure out how to make them run concurrently without affecting each other

Comment (rylev) [reply src/lib.rs:360]: Seems like a smell to have tests that read or write environment variables. However, I believe the encoding tests don't fall into this category, right? So for those tests, we should be able to get rid of the mutex. 

Perhaps tests that rely on reading and writing environment variables can be moved to using `spin-test` instead? In the meantime, we could remove the mutex for the tests that can be run in parallel and leave it for the other tests.

File: src/lib.rs [hunk 5]
```diff
@@ -49,47 +53,151 @@ const FALLBACK_FAVICON_ICO: &[u8] = include_bytes!("../spin-favicon.ico");
...
+        }
+    }
+}
+
+impl FromStr for ContentEncoding {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut parts = s.split(';');
+        let v = parts.next().unwrap().trim();
```
Comment (rylev) [inline src/lib.rs:112]: nit: I think we should prefer not using single character binding names in a somewhat large function like this:

```suggestion
        let encoding = parts.next().unwrap().trim();
```

File: src/lib.rs [hunk 6]
```diff
@@ -49,47 +53,151 @@ const FALLBACK_FAVICON_ICO: &[u8] = include_bytes!("../spin-favicon.ico");
...
+        let encoding = SupportedEncoding::from_str(v).context("failed to parse encoding type")?;
+        let Some(v) = parts
```
Comment (rylev) [inline src/lib.rs:114]:
```suggestion
        let Some(weight) = parts
```

**Review by rylev — APPROVED** [review]
Looks great! I think we should aim to add a `spin-test` based test as soon as https://github.com/fermyon/spin-fileserver/pull/55, but we don't need to necessarily block merging this PR on that one.

File: src/lib.rs (hunk 3 above)
Comment (endocrimes) [reply src/lib.rs:105]: Honestly a slightly more correct (not limiting to supported types, and supporting wildcards) implementation should probably be a crate of its own

File: src/lib.rs (hunk 1 above)
Comment (endocrimes) [reply src/lib.rs:360]: I've restricted the mutex to tests that interact with env - we should probably move them to spin-test tho later

File: src/lib.rs (hunk 4 above)
Comment (endocrimes) [reply src/lib.rs:186]: I think that _might_ be harder to read (or at least, would be for me), so without strong preference either way I'll leave it for now
